Add a new machine model for flexflow simulator

dycz0fx opened the pull request:
1. Restructured the simulator's code. Move the current code that describes the status of a machine into a separate class (MachineModel_old). This machine model would be used by default, and the simulator should behave the same as before.
2. Implemented a new machine model (MachineModel_new), which supports creating a machine model from a configure file (please see macine_config_example), and simulating message segmentation and possible congestions on a communication device.

lockshaw commented on include/simulator.h:
```diff
@@ -41,18 +42,178 @@ struct CostMetrics {
+    Comp_device(std::string name, CompDevType comp_type, int node_id, int socket_id, int device_id);
+};
+
+class Mem_device : public Device {
```
```suggestion
class MemDevice : public Device {
```

dycz0fx replied:
fixed

lockshaw commented on include/simulator.h:
```diff
@@ -41,18 +42,178 @@ struct CostMetrics {
+    int device_id;
+};
+
+class Comp_device : public Device {
```
```suggestion
class CompDevice : public Device {
```

dycz0fx replied:
fixed

lockshaw commented on include/simulator.h:
```diff
@@ -41,18 +42,178 @@ struct CostMetrics {
+  int version;
+};
+
+class MachineModel_old : public MachineModel {
```
```suggestion
class SimpleMachineModel : public MachineModel {
```

dycz0fx replied:
fixed

lockshaw commented on include/simulator.h:
```diff
@@ -41,18 +42,178 @@ struct CostMetrics {
+    CommDevType comm_type;
+    float latency;
+    float bandwidth;
+    Comm_device(std::string name, CommDevType comm_type, int node_id, int socket_id, int device_id, float latency, float bandwidth);
```
```suggestion
    Comm_device(std::string const &name, CommDevType comm_type, int node_id, int socket_id, int device_id, float latency, float bandwidth);
```

dycz0fx replied:
fixed

lockshaw commented on include/simulator.h:
```diff
@@ -41,18 +42,178 @@ struct CostMetrics {
+  float inter_gpu_bandwidth;
+  float inter_node_bandwidth;
+  float gpu_dram_bandwidth;
+  std::map<int, Comp_device*> id_to_gpu;
```
```suggestion
  std::unordered_map<int, Comp_device*> id_to_gpu;
```

dycz0fx replied:
I think it makes sense to change to unordered_map, but the original code uses map. Just trying to make the simple machine model be the same as the original one.

lockshaw commented on include/simulator.h:
```diff
@@ -41,18 +42,178 @@ struct CostMetrics {
+  std::map<size_t, Comm_device*> ids_to_inter_node_comm_device;
+};
+
+class MachineModel_new : public MachineModel {
```
```suggestion
class FullMachineModel : public MachineModel {
```

dycz0fx replied:
Maybe EnhancedMachineModel?

lockshaw commented on include/simulator.h:
```diff
@@ -41,18 +42,178 @@ struct CostMetrics {
+    std::vector<Comm_device::CommDevType> intra_socket_gpu_fb_mem_to_gpu_fb_mem;
+    std::vector<Comm_device::CommDevType> inter_socket_gpu_fb_mem_to_gpu_fb_mem;
+    std::vector<Comm_device::CommDevType> inter_node_gpu_fb_mem_to_gpu_fb_mem;
+    std::vector<std::vector<Comp_device *>> cpus;   // socket_id, local_id
```
Would it be reasonable to use a `std::unordered_map` with a struct key that makes these two indices clear, or is there a compelling reason for nested vectors?

lockshaw commented on include/simulator.h:
```diff
@@ -143,6 +305,7 @@ class TaskManager {
   SimTask* new_barrier_task();
   SimTask* new_update_task();
   SimTask* new_comm_task();
+  SimTask* new_comm_task(std::string name, Comm_device *comm_device, size_t message_size);
```
```suggestion
  SimTask* new_comm_task(std::string const &name, Comm_device *comm_device, size_t message_size);
```

dycz0fx replied:
fixed

lockshaw commented on include/simulator.h:
```diff
@@ -41,18 +42,178 @@ struct CostMetrics {
+
+class MachineModel_old : public MachineModel {
+public:
+  MachineModel_old(int num_nodes, int num_gpus_per_node, size_t capacity);
```
Feels like a lot of these methods could/should be `const`

dycz0fx replied:
fixed

lockshaw commented on include/simulator.h:
```diff
@@ -41,18 +42,178 @@ struct CostMetrics {
-  size_t capacity;
-  DeviceType type;
+    enum CompDevType {
+        LOC_PROC,   //CPU
```
Why are the names here unintuitive and require comments?

dycz0fx replied:
Just using Legion's name for the CPU and GPU.

lockshaw commented on src/runtime/simulator.cu:
```diff
@@ -114,8 +80,18 @@ void Simulator::strategy_search_task(const Task *task,
+    machine = (MachineModel *) new MachineModel_new(model->config.machine_model_file, gpu_mem.capacity());
+  }
+  else {
+    assert(false && "machine model creation error");
```
More useful error message here since this is not unlikely to happen?

dycz0fx replied:
Fixed.

lockshaw commented on src/runtime/simulator.cc:
```diff
@@ -39,29 +39,841 @@ bool ParallelConfig::is_data_parallel() const
+      nvlinks[i].push_back(new Comm_device(nvlink_name, Comm_device::NVLINK_COMM, node_id, socket_id, nvlink_id, latency, bandwidth));
+    }
+
+    for (int j = 0; j < num_sockets_per_node; j++) {
```
Can we use iterators or something to make this code a little easier to read?

lockshaw commented on src/runtime/simulator.cc:
```diff
@@ -39,29 +39,841 @@ bool ParallelConfig::is_data_parallel() const
+
+void MachineModel_new::add_pcis(float latency, float bandwidth)
+{
+  for (int i = 0; i < num_nodes; i++) {
```
```suggestion
  for (int node_id = 0; node_id < num_nodes; node_id++) {
```

lockshaw replied:
In general if we can substitute out most of the explicit indices with real names the code would be much easier to read

lockshaw commented on src/runtime/simulator.cc:
```diff
@@ -210,72 +1033,75 @@ void* Simulator::allocate(size_t num_elements, DataType type)
+  if (num_segment > 1 and path.size() >= 2) {
+    for (size_t i = 0; i < path.size(); i++) {
+      for (int j = 1; j < num_segment; j++) {
+        if (((Comm_device *)all_tasks[i][j]->device)->comm_type == Comm_device::NIC_OUT_COMM or
```
Divide this out into a couple expressions. It's quite hard to read

lockshaw commented on src/runtime/simulator.cc:
```diff
@@ -39,29 +39,841 @@ bool ParallelConfig::is_data_parallel() const
+            ret.emplace_back(id_to_dramtogpu_comm_device[tar_mem->device_id]);
+        }
+        else {
+            int device_id = src_mem->node_id * num_nodes + tar_mem->node_id;
```
Can we replace a bunch of these places where we do the hashing explicitly with instead creating structs and providing them `std::hash` instances? The inline hashing code is bug-prone and makes things hard to read

dycz0fx replied:
Make sense. When implementing the simple machine model, I tried to make it be the same as the original code, so the default behavior of Flexflow wouldn't be changed.

lockshaw commented on src/runtime/simulator.cc:
```diff
@@ -39,29 +39,841 @@ bool ParallelConfig::is_data_parallel() const
+        }
+    }
+    else {
+        printf("No path found between %s and %s\n", src_mem->name.c_str(), tar_mem->name.c_str());
```
Should this `exit` or `assert` or something, or is the expected behavior to just return an empty list?

dycz0fx replied:
fixed

lockshaw commented on src/runtime/simulator.cc:
```diff
@@ -39,29 +39,841 @@ bool ParallelConfig::is_data_parallel() const
+  num_gpus = num_sockets * num_gpus_per_socket;
+  num_nvlinks_per_node = 0;
+  mem_to_nvlink.clear();
+  add_cpus();
```
I'd personally prefer `this->` in front of these to make clear that they're methods and not functions, but this is a small thing. I defer to @jiazhihao on this

dycz0fx replied:
fixed

lockshaw reviewed (changes requested):
Overall it looks great, though I haven't so far been able to fully understand all of the improvements to the old simulator. I left a bunch of comments mostly focused on cleaning up some of the code and making it a bit easier to read. In general if if something was recurring I only commented once, so many of the suggestions should ideally be applied throughout the file. That said, I'm absolutely happy to discuss any suggestions you think are wrong :slightly_smiling_face:! Once you/we've worked through these, let me know and I'd love to give everything another read through before merging it. Thanks!

A couple comments that didn't make sense to put at a specific line of code: 
- Would it be possible to split some of the new machine model code out of simulator.cc? The file is getting quite large and it would be nice to have a clearer division between what code is related to the new machine model and what is related to the old one.
- It would also be great to have some doxygen-style docstrings provided for some of the more high-level functions in the new simulator, as well as a plain-text description summarizing roughly what the model includes and what it doesn't

dycz0fx commented on include/simulator.h:
```diff
@@ -41,18 +42,178 @@ struct CostMetrics {
+    void add_pcis(float latency, float bandwidth);
+    void add_nvlinks(float latency, float bandwidth);
+    void attach_nvlink(Mem_device *src_mem, Mem_device *tar_mem, Comm_device *comm);
+    void add_comm_path(std::vector<Comm_device::CommDevType> comm_device_list, Mem_device *src_mem, Mem_device *tar_mem, std::vector<Comm_device *> &ret);
```
Fixed

dycz0fx commented on src/runtime/simulator.cc:
```diff
@@ -39,29 +39,841 @@ bool ParallelConfig::is_data_parallel() const
+    switch (comm_device_list[i])
+    {
+    case Comm_device::MEMBUS_COMM:
+      ret.emplace_back(membuses[cur_mem->socket_id]);
```
I have an impression that emplace_back can save a copy as compared to push_back.

dycz0fx commented on src/runtime/simulator.cc:
```diff
@@ -39,29 +39,841 @@ bool ParallelConfig::is_data_parallel() const
+{
+  int hash = src_mem->device_id * num_gpus + tar_mem->device_id;
+  if (mem_to_nvlink.find(hash) != mem_to_nvlink.end()) {
+    return mem_to_nvlink[hash];
```
fixed

dycz0fx commented on src/runtime/simulator.cc:
```diff
@@ -115,7 +927,8 @@ SimTask* TaskManager::new_task()
   task->device = NULL;
-  task->op_name = NULL;
+  task->mem = NULL;
+  (task->name).clear();
```
fixed

dycz0fx commented on src/runtime/simulator.cc:
```diff
@@ -39,29 +39,841 @@ bool ParallelConfig::is_data_parallel() const
+      if (words.size() >= 3) {
+        if (words[0] == "num_nodes") {
+          num_nodes = stoi(words[2]);
+          printf("num_nodes = %d\n", num_nodes);
```
I used prinf because I saw the original code use it. What kind of logging mechanism would you suggest?

dycz0fx commented on src/runtime/simulator.cc:
```diff
@@ -39,29 +39,841 @@ bool ParallelConfig::is_data_parallel() const
-  bandwidth(0.0f),
-  capacity(_capacity),
-  type(_type)
+// class Device
```
fixed

dycz0fx commented on machine_config_example:
```diff
@@ -0,0 +1,49 @@
+# comm_device:
+# Communication devices describe the links between the memories. Each communication device needs two parameters - latency in ms and bandwidth in GB/s. An easy way to get these numbers is using the Memspeed benchmark in legion/test/realm. Currently, we provide the following communication devices:
+# memcpy 
+membus_latency = 0.00003
```
I used ms because the calculations in the simulator are in ms, but we can change to other units.

dycz0fx commented on src/runtime/simulator.cu:
```diff
@@ -114,8 +80,18 @@ void Simulator::strategy_search_task(const Task *task,
   // void* base_ptr = memFBImpl->get_direct_ptr(offset, 0);
+  MachineModel *machine;
+  if (model->config.machine_model_version == 0) {
+    machine = (MachineModel *) new MachineModel_old(model->config.numNodes, model->config.workersPerNode, gpu_mem.capacity());
```
fixed

dycz0fx commented on src/runtime/model.cc:
```diff
@@ -1924,6 +1924,9 @@ struct DefaultConfig {
   const static bool enableParameterParallel = false;
   const static bool enableAttributeParallel = false;
+  const static int machine_model_version = 0;
+  const static int simulator_segment_size = 16777216;
```
fixed

dycz0fx commented on src/runtime/simulator.cc:
```diff
@@ -39,29 +39,841 @@ bool ParallelConfig::is_data_parallel() const
+    // add_task_dependencies_with_xfer(src_task, dst_task, 64 << 20);
+}
+
+std::string MachineModel_new::to_string()
```
I did this way because I'd like to make this function virtual. It is a little bit trickier to make operator << virtual.

dycz0fx commented on machine_config_example:
```diff
@@ -0,0 +1,49 @@
+# This is an example of config file for the new machine model
```
I feel like the custom configuration file is pretty small and easy to handle. Using YMAL or XML is more standardized, but could make Flexflow require more external packages.

dycz0fx commented:
Thanks for reviewing the pull request! 
I addressed most of the comments and please take another look. Some comments about optimizing the old machine model make sense, but I feel like those should not be included in this pull request.
